platform/security/idp/security-idp-server/src/main/java/org/codice/ddf/security/idp/server/IdpEndpoint.java

```diff
@@ -228,37 +233,61 @@ public void init() {
         XMLObjectProviderRegistry xmlObjectProviderRegistry = ConfigurationService.get(
                 XMLObjectProviderRegistry.class);
         xmlObjectProviderRegistry.registerObjectProvider(Request.DEFAULT_ELEMENT_NAME,
-                new RequestBuilder(), new RequestMarshaller(), new RequestUnmarshaller());
-        xmlObjectProviderRegistry.registerObjectProvider(
-                ddf.security.liberty.paos.Response.DEFAULT_ELEMENT_NAME, new ResponseBuilder(),
-                new ResponseMarshaller(), new ResponseUnmarshaller());
+                new RequestBuilder(),
+                new RequestMarshaller(),
+                new RequestUnmarshaller());
+        xmlObjectProviderRegistry.registerObjectProvider(ddf.security.liberty.paos.Response.DEFAULT_ELEMENT_NAME,
+                new ResponseBuilder(),
+                new ResponseMarshaller(),
+                new ResponseUnmarshaller());
     }
 
-    private void parseServiceProviderMetadata(List<String> serviceProviderMetadata) {
-        if (serviceProviderMetadata != null) {
-            try {
-                MetadataConfigurationParser metadataConfigurationParser =
-                        new MetadataConfigurationParser(serviceProviderMetadata, ed -> {
-                            EntityInformation entityInfo = new EntityInformation.Builder(ed,
-                                    SUPPORTED_BINDINGS).build();
-                            if (entityInfo != null) {
-                                serviceProviders.put(ed.getEntityID(), entityInfo);
-                            }
-                        });
-
-                serviceProviders.putAll(metadataConfigurationParser.getEntryDescriptions()
-                        .entrySet()
-                        .stream()
-                        .map(e -> Maps.immutableEntry(e.getKey(),
-                                new EntityInformation.Builder(e.getValue(),
-                                        SUPPORTED_BINDINGS).build()))
-                        .filter(e -> nonNull(e.getValue()))
-                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
-
-            } catch (IOException e) {
-                LOGGER.warn("Unable to parse SP metadata configuration. Check the configuration for SP metadata.", e);
+    private Map<String, EntityInformation> getServiceProvidersMap() {
+        Map<String, EntityInformation> spMap = serviceProviders.get();
+        if (spMap == null) {
+            spMap = parseServiceProviderMetadata();
+
+            boolean updated = serviceProviders.compareAndSet(null, spMap);
+            if (!updated) {
+                spMap = serviceProviders.get();
             }
         }
+
+        return Collections.unmodifiableMap(spMap);
```

Comment:
Your analysis is definitely correct, but I'm not sure it matters in the real world. For this condition to occur, a thread would have to be updating the property concurrent with the first access. It isn't impossible, but the odds of it happening are vanishingly small. Setting the injected property is an action that is driven by administration of the system. Someone would need to be _resetting it_ while multiple users are logging in. Not initial set, but reset.

I can certainly catch the unlikely NPE and throw a different exception, but I think anything else is either overkill or is going to drag the system performance down.